File: Week_03/111/minimum_depth_of_binary_tree2.go
package main

func minDepth2(root *TreeNode) int {
	if root == nil {
		return 0
	}
	queue := []*TreeNode{root}
	dep := 0
	for len(queue) > 0 {
		dep++
		n := len(queue)
		for i := 0; i < n; i++ {
			if queue[i].Left == nil && queue[i].Right == nil {
				return dep
			}
			if queue[i].Left != nil {
				queue = append(queue, queue[i].Left)
			}
			if queue[i].Right != nil {
				queue = append(queue, queue[i].Right)
			}
		}
		queue = queue[n:]

	}
	return -1 // never reached
}

File: Week_03/111/minimum_depth_of_binary_tree2.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        dep = 0

        queue = [root]
        while queue:
            n = len(queue)
            dep += 1
            for i in range(n):
                if queue[i].left is None and queue[i].right is None:
                    return dep
                if queue[i].left is not None:
                    queue.append(queue[i].left)

                if queue[i].right is not None:
                    queue.append(queue[i].right)
            queue = queue[n:]
        return -1
